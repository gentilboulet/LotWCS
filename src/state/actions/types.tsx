import { IInitialStateAction } from './initial';
import { IHeaderAction } from './header';
import { IChiAction } from './chi';
import { ISkillAction } from './skills';
import { ILoresheetAction } from './loresheets';
import { IHistoryAction } from './history';
import { IKungFuAction } from './kungfus';
import { IVirtueAction } from './virtues';

export type IAction =
  IInitialStateAction | IHeaderAction | ISkillAction | IHistoryAction |
  ILoresheetAction | IChiAction | IKungFuAction | IVirtueAction;
