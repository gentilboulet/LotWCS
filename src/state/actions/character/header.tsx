import { createAction } from "typesafe-actions";

import * as dataArchetypes from "../../../data/archetypes";
import * as dataRanks from "../../../data/ranks";
import { IPerk } from "../../../perks";

export const setName = createAction("header/SET_NAME", (name: string) => {
  return { name };
})();

export const setConcept = createAction(
  "header/SET_CONCEPT",
  (concept: string) => {
    return { concept };
  },
)();

export const setArchetype = createAction(
  "header/SET_ARCHETYPE",
  (archetype: dataArchetypes.TArchetype) => {
    dataArchetypes.validateArchetype(archetype);
    return { archetype };
  },
)();

export const setRank = createAction(
  "header/SET_RANK",
  (rank: dataRanks.TRank) => {
    dataRanks.validateRank(rank);
    const dr = dataRanks.getRank(rank) as dataRanks.IDataRank;
    return { rank, perks: dr.perks as IPerk[] };
  },
)();
