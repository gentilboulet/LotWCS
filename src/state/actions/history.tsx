import * as constants from 'state/constants/history';

export interface IhistoryDeleteUpTo {
  type: constants.HISTORY_DELETE;
  id: number;
}

export type IHistoryAction = IhistoryDeleteUpTo;

export function historyDeleteUpTo(id: number): IhistoryDeleteUpTo {
  return { type: constants.HISTORY_DELETE, id };
}
