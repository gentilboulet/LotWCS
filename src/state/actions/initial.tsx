import * as constants from 'constants/initial';

export interface IInitialState {
  type: constants.INITIAL_STATE;
}

export type IInitialStateAction = IInitialState;

export function resetToInitialState(): IInitialState {
  return {type: constants.INITIAL_STATE};
}
