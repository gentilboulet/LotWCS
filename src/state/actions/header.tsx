import * as constants from 'state/constants/header';

import * as dataArchetypes from 'data/archetypes';
import * as dataRanks from 'data/ranks';

export interface IHeaderSetName {
  type: constants.HEADER_SET_NAME;
  name: string;
}

export interface IHeaderSetConcept {
  type: constants.HEADER_SET_CONCEPT;
  concept: string;
}

export interface IHeaderSetArchetype {
  type: constants.HEADER_SET_ARCHETYPE;
  archetype: string;
}

export interface IHeaderSetRank {
  type: constants.HEADER_SET_RANK;
  rank: string;
}

export type IHeaderAction = IHeaderSetName | IHeaderSetConcept | IHeaderSetArchetype | IHeaderSetRank;

export function setName(name: string): IHeaderSetName {
  return { type: constants.HEADER_SET_NAME, name };
}

export function setConcept(concept: string): IHeaderSetConcept {
  return { type: constants.HEADER_SET_CONCEPT, concept };
}

export function setArchetype(archetype: string): IHeaderSetArchetype {
  dataArchetypes.validateArchetype(archetype);
  return { archetype, type: constants.HEADER_SET_ARCHETYPE };
}

export function setRank(rank: string): IHeaderSetRank {
  dataRanks.validateRank(rank);
  return { rank, type: constants.HEADER_SET_RANK };
}
