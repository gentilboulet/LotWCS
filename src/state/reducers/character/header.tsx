import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/header";
import { ICharacterState } from "../../models/character";
import { setRank } from "../../models/character/header";

export const headerReducer = produce(
  (draft: Draft<ICharacterState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.setName):
        draft.name = action.payload.name;
        break;
      case getType(actions.setConcept):
        draft.concept = action.payload.concept;
        break;
      case getType(actions.setRank):
        setRank(draft, action.payload.rank);
        break;
      case getType(actions.setArchetype):
        draft.archetype = action.payload.archetype;
        break;
      default:
    }
  },
);
