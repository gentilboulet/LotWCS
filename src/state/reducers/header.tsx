import { IHeaderAction } from 'state/actions/header';
import * as constants from 'constants/header';
import * as dataRanks from 'data/ranks';
import { IStoreState } from 'state/types';

// Sub Reducers
import { IBonus } from 'perks/types/bonuses';
import { IDiscount } from 'perks/types/discounts';

import { applyBonuses } from 'perks/reducers/bonuses';
import { pushToDiscounts } from 'perks/reducers/discounts';
import { pushToHistory } from './history';

export function headerReducer(oldState: IStoreState, action: IHeaderAction): IStoreState {
  switch (action.type) {
    case constants.HEADER_SET_NAME:
      return oldState.withMutations(state => {
        state.set('name', action.name);
        pushToHistory(state, action);
      });
    case constants.HEADER_SET_CONCEPT:
      return oldState.withMutations(state => {
        state.set('concept', action.concept);
        pushToHistory(state, action);
      });
    case constants.HEADER_SET_RANK:
      return oldState.withMutations(state => {
        state.set('rank', action.rank );
        state.set('rankModified', true);

        const idx = dataRanks.ranks.findIndex((data) => (data.key === action.rank));
        state.set('rankValue', dataRanks.ranks[idx].value);

        applyBonuses(state, dataRanks.ranks[idx].perks as IBonus[]);
        pushToDiscounts(state, dataRanks.ranks[idx].perks as IDiscount[]);
        pushToHistory(state, action);
      });
    case constants.HEADER_SET_ARCHETYPE:
      return oldState.withMutations(state => {
        state.set('archetype', action.archetype );
        state.set('archetypeModified', true);
        pushToHistory(state, action);
      });
    default:
  }
  return oldState;
}
