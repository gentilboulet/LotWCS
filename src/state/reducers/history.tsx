import { IAction } from 'state/actions/types';
import { IStoreState } from 'state/types';

import { IHistoryAction } from 'state/actions/history';
import * as constants from 'state/constants/history';
import * as history from 'state/history';

export function historyReducer(oldState: IStoreState, action: IHistoryAction): IStoreState {
  switch (action.type) {
    case constants.HISTORY_DELETE:
      const list: IAction[] = oldState.get('history').toJS();
      return history.replayHistory(oldState, list.slice(0, action.id + 1)); // +1 needed to skip initialState
    default:
      return oldState;
  }
}
