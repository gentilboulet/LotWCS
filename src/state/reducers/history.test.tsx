import { IAction } from 'state/actions/types';
import { initialStateFactory } from 'state/initial';
import { IStoreState } from 'state/types';

import * as header from 'state/actions/header';
import { historyDeleteUpTo, IHistoryAction } from 'state/actions/history';
import { resetToInitialState } from 'state/actions/initial';
import * as history from 'state/history';

import { historyReducer } from './history';

const initialState: IStoreState = initialStateFactory();

describe('Testing historyReducer', () => {
  it('should receive an HISTORY_DELETE action', () => {
    const actions: IAction[] = [
      resetToInitialState(),
      header.setName('Roberts'),
      header.setConcept('Dread pirate Roberts'),
      header.setArchetype('warrior'),
      header.setRank('4th_rank'),
    ];
    const stateBefore = history.replayHistory(initialState, actions);
    expect( stateBefore ).toMatchSnapshot();
    const state = history.replayHistory(stateBefore, [historyDeleteUpTo(1)]);
    expect( state ).toMatchSnapshot();
    expect( state.get('name') ).toBe('Roberts');
    expect( state.get('concept') ).toBe('No Concept');
    expect( state.get('history').length === (1 + 1) );
  });

  it('should do nothing with a junk action', () => {
    const junk = { type: 'JUNK_ACTION' };
    expect( historyReducer(initialState, junk as IHistoryAction )).toMatchSnapshot();
  });
});
