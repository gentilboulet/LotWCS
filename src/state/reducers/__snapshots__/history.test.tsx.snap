// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing historyReducer should do nothing with a junk action 1`] = `
Object {
  "archetype": undefined,
  "bonuses": Array [],
  "chi": Object {
    "corrupt": Object {
      "cultivation": 0,
      "value": 0,
    },
    "earth": Object {
      "cultivation": 0,
      "value": 0,
    },
    "enlightened": Object {
      "cultivation": 0,
      "value": 0,
    },
    "fire": Object {
      "cultivation": 0,
      "value": 0,
    },
    "general": Object {
      "cultivation": 0,
      "value": 0,
    },
    "metal": Object {
      "cultivation": 0,
      "value": 0,
    },
    "water": Object {
      "cultivation": 0,
      "value": 0,
    },
    "wood": Object {
      "cultivation": 0,
      "value": 0,
    },
  },
  "concept": undefined,
  "destiny": 0,
  "discounts": Array [],
  "entanglement": 0,
  "history": Array [
    Object {
      "type": "INITIAL_STATE",
    },
  ],
  "kungfu": Object {
    "KUNGFU_EXTERNAL": Object {},
    "KUNGFU_INTERNAL": Object {},
  },
  "loresheets": Object {},
  "name": undefined,
  "rank": undefined,
  "skills": Object {
    "Awareness": Object {
      "name": "Awareness",
      "specialities": Array [],
      "value": 0,
    },
    "Confidence": Object {
      "name": "Confidence",
      "specialities": Array [],
      "value": 0,
    },
    "Crafting": Object {
      "name": "Crafting",
      "specialities": Array [],
      "value": 0,
    },
    "Finesse": Object {
      "name": "Finesse",
      "specialities": Array [],
      "value": 0,
    },
    "Hardiness": Object {
      "name": "Hardiness",
      "specialities": Array [],
      "value": 0,
    },
    "Inspire": Object {
      "name": "Inspire",
      "specialities": Array [],
      "value": 0,
    },
    "Learning": Object {
      "name": "Learning",
      "specialities": Array [],
      "value": 0,
    },
    "Medicine": Object {
      "name": "Medicine",
      "specialities": Array [],
      "value": 0,
    },
    "Might": Object {
      "name": "Might",
      "specialities": Array [],
      "value": 0,
    },
    "Perform": Object {
      "name": "Perform",
      "specialities": Array [],
      "value": 0,
    },
    "Politics": Object {
      "name": "Politics",
      "specialities": Array [],
      "value": 0,
    },
    "Ride": Object {
      "name": "Ride",
      "specialities": Array [],
      "value": 0,
    },
    "Stealth": Object {
      "name": "Stealth",
      "specialities": Array [],
      "value": 0,
    },
    "Survival": Object {
      "name": "Survival",
      "specialities": Array [],
      "value": 0,
    },
    "Tactics": Object {
      "name": "Tactics",
      "specialities": Array [],
      "value": 0,
    },
    "Wu Wei": Object {
      "name": "Wu Wei",
      "specialities": Array [],
      "value": 0,
    },
  },
  "virtues": Array [
    Object {
      "name": "Honor",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Benevolence",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Righteousness",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Loyalty",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Force",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Revenge",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Individualism",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Obsession",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ruthlessness",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ferocity",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
  ],
}
`;

exports[`Testing historyReducer should receive an HISTORY_DELETE action 1`] = `
Object {
  "archetype": "warrior",
  "bonuses": Array [],
  "chi": Object {
    "corrupt": Object {
      "cultivation": 0,
      "value": 0,
    },
    "earth": Object {
      "cultivation": 0,
      "value": 0,
    },
    "enlightened": Object {
      "cultivation": 0,
      "value": 0,
    },
    "fire": Object {
      "cultivation": 0,
      "value": 0,
    },
    "general": Object {
      "cultivation": 0,
      "value": 10,
    },
    "metal": Object {
      "cultivation": 0,
      "value": 0,
    },
    "water": Object {
      "cultivation": 0,
      "value": 0,
    },
    "wood": Object {
      "cultivation": 0,
      "value": 0,
    },
  },
  "concept": "Dread pirate Roberts",
  "destiny": 20,
  "discounts": Array [
    Object {
      "skills": Array [
        "Awareness",
        "Confidence",
        "Crafting",
        "Finesse",
        "Hardiness",
        "Inspire",
        "Learning",
        "Medicine",
        "Might",
        "Perform",
        "Politics",
        "Ride",
        "Stealth",
        "Survival",
        "Tactics",
        "Wu Wei",
      ],
      "type": "DISCOUNT_SKILL",
      "value": 20,
    },
  ],
  "entanglement": 0,
  "history": Array [
    Object {
      "type": "INITIAL_STATE",
    },
    Object {
      "name": "Roberts",
      "type": "HEADER_SET_NAME",
    },
    Object {
      "concept": "Dread pirate Roberts",
      "type": "HEADER_SET_CONCEPT",
    },
    Object {
      "archetype": "warrior",
      "type": "HEADER_SET_ARCHETYPE",
    },
    Object {
      "rank": "4th_rank",
      "type": "HEADER_SET_RANK",
    },
  ],
  "kungfu": Object {
    "KUNGFU_EXTERNAL": Object {},
    "KUNGFU_INTERNAL": Object {},
  },
  "loresheets": Object {},
  "name": "Roberts",
  "rank": Object {
    "name": "4th_rank",
    "value": 2,
  },
  "skills": Object {
    "Awareness": Object {
      "name": "Awareness",
      "specialities": Array [],
      "value": 0,
    },
    "Confidence": Object {
      "name": "Confidence",
      "specialities": Array [],
      "value": 0,
    },
    "Crafting": Object {
      "name": "Crafting",
      "specialities": Array [],
      "value": 0,
    },
    "Finesse": Object {
      "name": "Finesse",
      "specialities": Array [],
      "value": 0,
    },
    "Hardiness": Object {
      "name": "Hardiness",
      "specialities": Array [],
      "value": 0,
    },
    "Inspire": Object {
      "name": "Inspire",
      "specialities": Array [],
      "value": 0,
    },
    "Learning": Object {
      "name": "Learning",
      "specialities": Array [],
      "value": 0,
    },
    "Medicine": Object {
      "name": "Medicine",
      "specialities": Array [],
      "value": 0,
    },
    "Might": Object {
      "name": "Might",
      "specialities": Array [],
      "value": 0,
    },
    "Perform": Object {
      "name": "Perform",
      "specialities": Array [],
      "value": 0,
    },
    "Politics": Object {
      "name": "Politics",
      "specialities": Array [],
      "value": 0,
    },
    "Ride": Object {
      "name": "Ride",
      "specialities": Array [],
      "value": 0,
    },
    "Stealth": Object {
      "name": "Stealth",
      "specialities": Array [],
      "value": 0,
    },
    "Survival": Object {
      "name": "Survival",
      "specialities": Array [],
      "value": 0,
    },
    "Tactics": Object {
      "name": "Tactics",
      "specialities": Array [],
      "value": 0,
    },
    "Wu Wei": Object {
      "name": "Wu Wei",
      "specialities": Array [],
      "value": 0,
    },
  },
  "virtues": Array [
    Object {
      "name": "Honor",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Benevolence",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Righteousness",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Loyalty",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Force",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Revenge",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Individualism",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Obsession",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ruthlessness",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ferocity",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
  ],
}
`;

exports[`Testing historyReducer should receive an HISTORY_DELETE action 2`] = `
Object {
  "archetype": undefined,
  "bonuses": Array [],
  "chi": Object {
    "corrupt": Object {
      "cultivation": 0,
      "value": 0,
    },
    "earth": Object {
      "cultivation": 0,
      "value": 0,
    },
    "enlightened": Object {
      "cultivation": 0,
      "value": 0,
    },
    "fire": Object {
      "cultivation": 0,
      "value": 0,
    },
    "general": Object {
      "cultivation": 0,
      "value": 0,
    },
    "metal": Object {
      "cultivation": 0,
      "value": 0,
    },
    "water": Object {
      "cultivation": 0,
      "value": 0,
    },
    "wood": Object {
      "cultivation": 0,
      "value": 0,
    },
  },
  "concept": undefined,
  "destiny": 0,
  "discounts": Array [],
  "entanglement": 0,
  "history": Array [
    Object {
      "type": "INITIAL_STATE",
    },
    Object {
      "name": "Roberts",
      "type": "HEADER_SET_NAME",
    },
  ],
  "kungfu": Object {
    "KUNGFU_EXTERNAL": Object {},
    "KUNGFU_INTERNAL": Object {},
  },
  "loresheets": Object {},
  "name": "Roberts",
  "rank": undefined,
  "skills": Object {
    "Awareness": Object {
      "name": "Awareness",
      "specialities": Array [],
      "value": 0,
    },
    "Confidence": Object {
      "name": "Confidence",
      "specialities": Array [],
      "value": 0,
    },
    "Crafting": Object {
      "name": "Crafting",
      "specialities": Array [],
      "value": 0,
    },
    "Finesse": Object {
      "name": "Finesse",
      "specialities": Array [],
      "value": 0,
    },
    "Hardiness": Object {
      "name": "Hardiness",
      "specialities": Array [],
      "value": 0,
    },
    "Inspire": Object {
      "name": "Inspire",
      "specialities": Array [],
      "value": 0,
    },
    "Learning": Object {
      "name": "Learning",
      "specialities": Array [],
      "value": 0,
    },
    "Medicine": Object {
      "name": "Medicine",
      "specialities": Array [],
      "value": 0,
    },
    "Might": Object {
      "name": "Might",
      "specialities": Array [],
      "value": 0,
    },
    "Perform": Object {
      "name": "Perform",
      "specialities": Array [],
      "value": 0,
    },
    "Politics": Object {
      "name": "Politics",
      "specialities": Array [],
      "value": 0,
    },
    "Ride": Object {
      "name": "Ride",
      "specialities": Array [],
      "value": 0,
    },
    "Stealth": Object {
      "name": "Stealth",
      "specialities": Array [],
      "value": 0,
    },
    "Survival": Object {
      "name": "Survival",
      "specialities": Array [],
      "value": 0,
    },
    "Tactics": Object {
      "name": "Tactics",
      "specialities": Array [],
      "value": 0,
    },
    "Wu Wei": Object {
      "name": "Wu Wei",
      "specialities": Array [],
      "value": 0,
    },
  },
  "virtues": Array [
    Object {
      "name": "Honor",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Benevolence",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Righteousness",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Loyalty",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Force",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Revenge",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Individualism",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Obsession",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ruthlessness",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ferocity",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
  ],
}
`;
